rfcs/20200612-stream-executor-c-api.md

```diff
@@ -308,26 +308,71 @@ typedef struct SE_StreamExecutor {
 
 #define SE_STREAMEXECUTOR_STRUCT_SIZE TF_OFFSET_OF_END(SE_StreamExecutor, host_callback)
 
-TF_CAPI_EXPORT SE_Platform* SE_NewPlatform(
-     SE_PlatformId* id,
-     int32_t visible_device_count,
-     SE_Device* (*create_device)(SE_Options* options, TF_Status* status),
-     void (*destroy_device)(SE_Device* device),
-     SE_StreamExecutor* (*create_stream_executor)(TF_Status* status),
-     void (*destroy_stream_executor)(SE_StreamExecutor* stream_executor);
-);
+typedef struct SE_PlatformParams {
+  size_t struct_size;
+  SE_PlatformId* id;
+  
+  // Callbacks for creating/destroying SE_Device.
+  SE_Device* (*create_device)(SE_Options* options, TF_Status* status);
```

Comment:
1. This API implies that SE_Device must created on the heap, and returns a pointer to the allocated memory. This typically makes sense especially for opaque types.

In our case, SE_Device is visible to TF and the plugin, we should allow TF to create SE_device on either heap or stack and passing that to the plugin for initialization.

2. More importantly, IIUC the use case is for TF to request from the plugin an SE_Device to use, based on its knowledge of what SE_Device looks like:

> TF: Here's what I think SE_Device looks like, please fill it in as much as possible.

In other words, TF is compiled against SEI and a certain layout of SE_Device which it _expects_, and doesn't make sense for the plugin to guess what TF's view of SE_Device is.

Proposed:
```
// API
void (*init_device)(
  SE_Options* options,  // in
  SE_Device** device,   // out
  TF_Status* status     // out
);

// TF code
void TF_InitDevice(const SE_PlatformParams& platform, const SE_Options& options) {
  TF_Status status;

  // SE_Device on stack
  SE_Device device_stack{ SE_DEVICE_STRUCT_SIZE };
  platform.init_device(options, &device_stack, &status);

  // SE_Device on heap
  SE_Device* device_heap = new SE_Device{ SE_DEVICE_STRUCT_SIZE };
  platform.init_device(options, device_heap, &status);

}
```
The way a plugin can populate TF's SE_Device is by taking a hint from struct_size:
-   A newer plugin will populate the provided SE_Device up to the provided device_from_tf->struct_size.
-   An older plugin will populate the provided SE_Device up to the plugin's (smaller) knowledge of SE_Device_STRUCT_SIZE.


Sorry for the long reply, I hope this makes sense!

Reply:
Did you mean

```
SE_Device* device,   // out
```
?

Reply:
Good eye, thanks! :100:  That was indeed a typo.

Reply:
Thank you for a useful observation and detailed examples! Updated.